README.md
# node-realtime-chat-app-example

An example realtime chat application using Node.js and [socket.io](http://socket.io/)
```
git clone https://github.com/TeamworkGuy2/node-realtime-chat-app-example.git
cd node-realtime-chat-app-example/
npm install
npm start
```
And just open localhost:3000 in your browser

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project does its best to adhere to [Semantic Versioning](http://semver.org/).


--------
### [0.2.3](N/A) - 2017-04-29
#### Added
* test/sql-query-example.js - contains an example script to query a SQL DB


--------
### [0.2.2](https://github.com/TeamworkGuy2/node-realtime-chat-app-example/commit/16dbaaa5f24b65af7661e9f12255734588a0df71) - 2017-04-24
#### Added
* server/server-with-jokes.js - containing full server code ready to show


--------
### [0.2.1](https://github.com/TeamworkGuy2/node-realtime-chat-app-example/commit/6ccc81d8822ee67667830f39f0b86b93b871b7c4) - 2016-07-26
#### Changed
* Fixed some indentation, added some documentation, and prepped for knock-knock joke version of server.js


--------
### [0.2.0](https://github.com/TeamworkGuy2/node-realtime-chat-app-example/commit/a201b4228586e62570572daf28233bd8889b5fa8) - 2016-07-26
#### Changed
* Refactored message format and types of messages (now includes 'query').
* Changed 'when' property from Date to timestamp


--------
### [0.1.2](https://github.com/TeamworkGuy2/node-realtime-chat-app-example/commit/58cb13ea7dc6e4bcb4c1ac47486528cdfcdfc5d8) - 2016-07-20
#### Added
Setup server.js for easier administration


--------
### [0.1.1](https://github.com/TeamworkGuy2/node-realtime-chat-app-example/commit/94b332cf32113316310d223650d512de0a5770d5) - 2016-07-19
#### Added
README.md and CHANGELOG.md

#### Fixed
Some path issues with server/server.js


--------
### [0.1.0](https://github.com/TeamworkGuy2/node-realtime-chat-app-example/commit/c02e93437afee290be71efcd6520549436dee7c1) - 2016-07-19
#### Added
Initial fork of original chat app

#### Changed
Should work out of the box with node.js, npm, and no other requirements

#### Removed
'redis' package requirement in favor of in-memory message storage (lost when server restarts) for now


test/sql-query-example.js
var sql = require("mssql");
 
sql.connect("mssql://user_id:password@server_name/db_name").then(function() {
    // Query 
	new sql.Request().query('select * from table_name').then(function(recordset) {
		console.dir(recordset);
	}).catch(function(err) {
		console.error("query error", err);
	});

    // Stored Procedure 
	/*new sql.Request()
	.input('input_parameter', sql.Int, value)
    .output('output_parameter', sql.VarChar(50))
	.execute('procedure_name').then(function(recordsets) {
		console.dir(recordsets);
	}).catch(function(err) {
		// ... execute error checks 
	});*/

}).catch(function(err) {
	console.error("connection error", err);
});


server/server.js
var fs = require("fs");
var express = require("express");
var http = require("http");
var path = require("path");
var socketIo = require("socket.io");
var db = require("./in-mem-db");

var webPort = 3000;
var socketPort = 53000;
var httpServer = http.Server(app);
var ioSocket = socketIo(httpServer);
var app = express();

httpServer.listen(process.env.PORT || socketPort);


// Connect to DB
var dbConn;
var msgsCollection;

function createDb() {
  dbConn = db.newInMemDb({
    name: "hds-devcon-chat-app",
    collectionConfigs: [{
      name: "messages",
      maxSize: 100
    }]
  });

  msgsCollection = dbConn.getCollection("messages");
}

createDb();

function getSortedItems() {
  return msgsCollection.getItems().sort(function (a, b) {
    return a.when < b.when ? 1 : a.when > b.when ? -1: 0;
  });
}

function copyCollection(items) {
  return JSON.parse(JSON.stringify(items));
}


// Configure Jade template engine
app.set("views", path.join(__dirname, "../views"));
app.set("view engine", "jade");
app.use(express.static(path.join(__dirname + "/../", "public")));
app.use("/socket.io", express.static(path.join(__dirname, "../node_modules/socket.io-client/")));

// handle HTTP GET request to the "/" URL
app.get("/", function (req, res) {

  console.log("requesting: " + req.method + " " + req.url);

  // Get the 100 most recent messages from the DB
  var items = copyCollection(getSortedItems());
  items.forEach(function (i) { i.isPost = i.action == "user-posted-message"; });

  // Pass the message list to the view
  res.render("index", { messages: items });    
});

app.listen(webPort, function () {
  console.log("starting chat app on port " + webPort);
});


var users = [];

// socket.io connection to listen for messages
ioSocket.on("connection", function (socket) {

  // When a message is received, broadcast it 
  // to all users except the originating client
  socket.on("msg", function (msgData) {
    console.log("message: ", msgData);

    msgData.action = "user-posted-message";
    msgData.isPost = true;
    msgsCollection.addItem(msgData);

    socket.broadcast.emit("msg", msgData);
  });


  // When a user joins the chat, send a notice
  // to all users except the originating client
  socket.on("join", function (joinData) {
    var nickname = joinData.nickname;
    users.push(socket);

    console.log("join: ", nickname, "(" + users.length + " users)");

    joinData.action = "user-connected";
    joinData.msg = nickname + " has joined the chat.";
    msgsCollection.addItem(joinData);
    // Attach the user's nickname to the socket
    socket.nickname = nickname;
    socket.broadcast.emit("notice", joinData);
    socket.emit("notice", joinData);
  });


  // When a user renames themself, send a notice
  // to all users except the originating client
  socket.on("rename", function (renameData) {
    var oldName = renameData.oldName;
    var newName = renameData.newName;

    console.log("rename: ", oldName, "to", newName);

    renameData.action = "user-renamed";
    renameData.msg = oldName + " has renamed themself " + newName;
    msgsCollection.addItem(renameData);
    // Attach the user's nickname to the socket
    socket.nickname = newName;
    socket.broadcast.emit("notice", renameData);
  });


  // 'query' expects an object with min and/or max timestamps and sends back
  // an array of messages which have a timestamp falling within that range
  socket.on("query", function (queryData) {
    var min  = queryData.minTimestamp || Number.MIN_SAFE_INTEGER;
    var max  = queryData.maxTimestamp || Number.MAX_SAFE_INTEGER;

    var items = getSortedItems();
    var res = [];

    for(var i = 0, size = items.length; i < size; i++) {
      var item = items[i];
      if(min < item.when && max > item.when) {
        res.push(item);
      }
    }

    console.log("query min: " + min + ", max: " + max + ", " + res.length + " matches of " + items.length);

    socket.emit("query", res);
  });


  // When a user disconnects, send a notice
  // to all users except the originating client
  socket.on("disconnect", function () {
    remove(users, socket);

    console.log("leave: ", socket.nickname, "(" + users.length + " users)");

    var leaveData = {
      action: "user-disconnect",
      msg: socket.nickname + " has left the chat.",
      nickname: socket.nickname,
      when: Date.now()
    };
    msgsCollection.addItem(leaveData);

    socket.broadcast.emit("notice", leaveData);
  });

});


// ==== utils ====

/** remove the first matching element from an array */
function remove(ary, val) {
  for (var i = 0, size = ary.length; i < size; i++) {
    if (ary[i] === val) {
      ary[i] = ary[size - 1];
      ary.pop();
      return true;
    }
  }
  return false;
}


/** emit a message to all listening sockets */
function emitToAll(type, data) {
  var clients = ioSocket.sockets.sockets;
  for (var i = 0, size = clients.length; i < size; i++) {
    clients[i].emit(type, data);
  }
}


/** run a task at a random time between now and maxIntervalMs and then wait for the task to return before setting up another timer to run it again and again... */
function setupAwaitTaskTimer(maxIntervalMs, task) {
  setTimeout(function () {
    var c = 0;
    task(function () {
      c++;
      if (c === 1) {
        setupAwaitTaskTimer(maxIntervalMs, task);
      }
    });
  }, Math.random() * maxIntervalMs);
}


server/in-mem-db.js

// ==== InMemDb ====
/** Create a simple in-memory database
 * @param opts: contains the following optional properties
 * [name] optional name of the collection, else name will be generated based on timestamp
 * [collectionConfigs] an array of collection configuration options, see addCollection()
 */
function InMemDb(opts) {
	opts = opts || {};

	this.name = opts.name || ("in-mem-db-" + Date.now());
	this.collections = [];

	if(opts.collectionConfigs && Array.isArray(opts.collectionConfigs)) {
		for(var i = 0, size = opts.collectionConfigs.length; i < size; i++) {
			this.addCollection(opts.collectionConfigs[i]);
		}
	}
}

/** Add a collection with various configuration options.
 * @param opts: contains options configuring the collection to add, including the following properties:
 * {string} name the collection's name
 * {int} [maxSize] the maximum size of the collection (default: Infinity)
 */
InMemDb.prototype.addCollection = function addCollection(opts) {
	if(opts == null) { throw new Error("cannot create collection without config options"); }
	var newCollection = null;
	if((newCollection = this.getCollection(opts.name)) != null) { return newCollection; }

	newCollection = new Collection({
		name: opts.name,
		maxSize: opts.maxSize || Infinity,
		data: []
	});

	this.collections.push(newCollection);

	return newCollection;
};


/** Get a collection by name
 * @param name the name of the collection to retrieve
 * @return this database's collection matching the given name, or null if no such collection exists
 */
InMemDb.prototype.getCollection = function getCollection(name) {
	var colls = this.collections;
	for(var i = 0, size = colls.length; i < size; i++) {
		if(colls[i].name === name) { return colls[i]; }
	}
	return null;
};


/** Remove a collection by name
 * @return true if the collection was found and removed, false if not
 */
InMemDb.prototype.removeCollection = function removeCollection(name) {
	var colls = this.collections;
	for(var i = 0, size = colls.length; i < size; i++) {
		if(colls[i].name === name) {
			colls[i] = colls[size - 1];
			colls.pop();
			return true;
		}
	}
	return false;
};




// ==== Collection ====
/**
 * @param opts: contains options configuring the collection to add, including the following properties:
 * {string} name the collection's name
 * {int} maxSize the maximum size of the collection (default: Infinity)
 * {Array} data the data array to store the collection's items in
 */
function Collection(opts) {
	this.name = opts.name;
	this.maxSize = opts.maxSize;
	this.data = opts.data;
}

Collection.prototype.addItem = function addItem(item) {
	if(item == null) { return; }
	this.data.push(item);

	if(this.maxSize != null && this.data.length > this.maxSize) {
		this.data.shift();
	}
};

Collection.prototype.addItems = function addItems(items) {
	if(items == null && Array.isArray(items)) { return; }
	Array.prototype.push.apply(this, items);

	if(this.maxSize != null && this.data.length > this.maxSize) {
		this.data.splice(0, this.maxSize - this.data.length);
	}
};

Collection.prototype.clear = function clear() {
	this.data = [];
};

/**
 * @param {function(T,int,T[])|int} offsetOrFilter: either a function (which is passed parameters mirroring Array.filter's callback)
 * or number, in which case it is used as an offset into this collection at which to slice off a sub-array of items to return
 * {int} [len] optional number of items to return, only used if the first argument 'offsetOrFilter' is a number
 */
Collection.prototype.getItems = function getItems(offsetOrFilter, len) {
	var res = [];
	var dat = this.data;
	if(typeof offsetOrFilter === "function") {
		var filter = offsetOrFilter;
		for(var i = 0, size = dat.length; i < size; i++) {
			if(!!filter(dat[i], i, dat)) {
				res.push(dat[i]);
			}
		}
	}
	else {
		var off = offsetOrFilter;
		res = dat.slice(off, len != null ? off + len : undefined);
	}
	return res;
};



// ==== exports ====
exports.newInMemDb = function (opts) {
	return new InMemDb(opts || {});
};

public/javascripts/client.js
$(document).ready(function onLoad() {
  var socketPort = 53000;
  var path = window.location.protocol + "//" + window.location.hostname + ":" + socketPort;
  var socket = io(path);
  var msgListElem = $('#messages');
  var user = { nickname: null };

  // Check if nickname stored in localStorage
  if('localStorage' in window && localStorage.getItem('nickname')) {
    user.nickname = localStorage.getItem('nickname');
  }
  else {
    user.nickname = promptAndSaveNickname();
  }

  setNickname(user.nickname);
  setupNicknameChangeListener(socket, user, msgListElem);

  // Send message to server that user has joined
  socket.emit('join', {
    nickname: user.nickname,
    when: Date.now()
  });

  // Handle the form to submit a new message
  $('form').submit(function submitForm(e) {
    e.preventDefault();

    var msgField = $('#msg');
    var msgVal = msgField.val();
    if(msgVal.trim().length === 0) { return; }

    var data = {
      msg: msgVal,
      nickname: user.nickname,
      when: Date.now()
    };

    // Send message to Socket.io server
    socket.emit('msg', data);
    // Add message to the page
    addMessage(data, msgListElem);
    // Clear the message field
    msgField.val('');    
  });

  var alsoAdd = [];

  // When a message is received from the server add it to the page
  socket.on('msg', function (data) {
    if(firstMsg) {
      alsoAdd.push(data);
      queryForMissingMsgs(socket, data, alsoAdd, msgListElem);
    }
    else {
      addMessage(data, msgListElem);
    }
  });

  // When a notice is received from the server (user renamed, joins, or leaves), add it to the page
  socket.on('notice', function (data) {
    if(firstMsg) {
      alsoAdd.push(data);
      queryForMissingMsgs(socket, data, alsoAdd, msgListElem);
    }
    else {
      addNotice(data, msgListElem);
    }
  });
});


function promptAndSaveNickname(existingName) {
  // If not in localStorage, prompt user for nickname
  var nickname = prompt('Please enter your nickname', existingName);
  if('localStorage' in window) {
    localStorage.setItem('nickname', nickname);
  }
  return nickname;
}


var firstMsg = true;

function queryForMissingMsgs(socket, latestPost, alsoAdd, dstList) {
  if(!firstMsg) { return; }

  var min = getLastMsgTimestampFromUi(dstList);
  var max = latestPost.when;
  socket.emit("query", { minTimestamp: min, maxTimestamp: max, when: Date.now() });

  socket.on("query", function (posts) {
    if(!firstMsg) { return; }

    firstMsg = false;
    addPosts(posts, dstList);
    addPosts(alsoAdd, dstList);
  });
}


function addPosts(posts, dstList) {
  for(var i = 0, size = posts.length; i < size; i++) {
    var post = posts[i];
    if(post.isPost) {
      addMessage(post, dstList);
    }
    else {
      addNotice(post, dstList);
    }
  }
}


function getLastMsgTimestampFromUi(msgList) {
  var lastMsg = msgList.find("li, div")[0];
  return lastMsg ? parseInt(lastMsg.dataset.postDate) : null;
}


function setNickname(nameStr) {
  $('.user-name').text("user: " + nameStr);
}


function setupNicknameChangeListener(socket, user, dstList) {
  $('a.user-name').on("click", function () {
    var newName = promptAndSaveNickname(user.nickname);
    var oldName = user.nickname;
    if(newName !== user.nickname) {
      socket.emit("rename", {
        oldName: oldName,
        newName: newName,
        when: Date.now()
      });

      setNickname(newName);
      user.nickname = newName;
      addNotice({ msg: oldName + " has renamed themself " + newName }, dstList);
    }
  });
}


// Function to add a notice (user renamed, joins, or leaves) to the page
function addNotice(data, dstList) {
    var now = Date.now();
    var notice = $('<div class="notice">').text(data.msg);
    notice[0].dataset.postDate = now;

    dstList.prepend(notice);
}


// Function to add a message to the page
function addMessage(data, dstList) {
  var now = Date.now();
  var who = $('<div class="who">').text(data.nickname);
  var when = $('<div class="when">').text(new Date(now).toString().substr(0, 24));
  var msg = $('<div class="msg">').text(data.msg);
  var header = $('<div class="header clearfix">').append(who).append(when);
  var li = $('<li>').append(header).append(msg);
  li[0].dataset.postDate = now;

  dstList.prepend(li);
}


server/server-with-jokes.js
var fs = require("fs");
var express = require("express");
var http = require("http");
var path = require("path");
var socketIo = require("socket.io");
var db = require("./in-mem-db");

var webPort = 3000;
var socketPort = 53000;
var httpServer = http.Server(app);
var ioSocket = socketIo(httpServer);
var app = express();

var jokeIntervalMs = 10000;
var jokes = JSON.parse(fs.readFileSync("./server/knock-knock-jokes.json", "UTF-8"));

httpServer.listen(process.env.PORT || socketPort);


// Connect to DB
var dbConn;
var msgsCollection;

function createDb() {
  dbConn = db.newInMemDb({
    name: "hds-devcon-chat-app",
    collectionConfigs: [{
      name: "messages",
      maxSize: 100
    }]
  });

  msgsCollection = dbConn.getCollection("messages");
}

createDb();

function getSortedItems() {
  return msgsCollection.getItems().sort(function (a, b) {
    return a.when < b.when ? 1 : a.when > b.when ? -1: 0;
  });
}

function copyCollection(items) {
  return JSON.parse(JSON.stringify(items));
}


// Configure Jade template engine
app.set("views", path.join(__dirname, "../views"));
app.set("view engine", "jade");
app.use(express.static(path.join(__dirname + "/../", "public")));
app.use("/socket.io", express.static(path.join(__dirname, "../node_modules/socket.io-client/")));

// handle HTTP GET request to the "/" URL
app.get("/", function (req, res) {

  console.log("requesting: " + req.method + " " + req.url);

  // Get the 100 most recent messages from the DB
  var items = copyCollection(getSortedItems());
  items.forEach(function (i) { i.isPost = i.action == "user-posted-message"; });

  // Pass the message list to the view
  res.render("index", { messages: items });    
});

app.listen(webPort, function () {
  console.log("starting chat app on port " + webPort);
});


var users = [];

// socket.io connection to listen for messages
ioSocket.on("connection", function (socket) {

  // When a message is received, broadcast it 
  // to all users except the originating client
  socket.on("msg", function (msgData) {
    console.log("message: ", msgData);

    var msgL = msgData.msg.toLowerCase();

    if (jokeState == "knock" && msgL === "who's there?") {
      jokeState = "who-there";
      setTimeout(function () {
        var msg = { msg: curJoke.who, when: Date.now() };
        emitToAll("notice", msg);
        msgsCollection.addItem(msg);
      }, 50);
    }
    else if (jokeState == "who-there" && msgL === (curJoke.who + " who").toLowerCase()) {
      jokeState = "who";
      setTimeout(function () {
        jokeState = null;
        var msg = { msg: curJoke.punchline, when: Date.now() };
        emitToAll("notice", msg);
        msgsCollection.addItem(msg);
        if (curCb) { curCb(); }
      }, 50);
    }

    msgData.action = "user-posted-message";
    msgData.isPost = true;
    msgsCollection.addItem(msgData);

    socket.broadcast.emit("msg", msgData);
  });


  // When a user joins the chat, send a notice
  // to all users except the originating client
  socket.on("join", function (joinData) {
    var nickname = joinData.nickname;
    users.push(socket);

    console.log("join: ", nickname, "(" + users.length + " users)");

    if (jokeCount < 0) {
      postJoke();
    }

    joinData.action = "user-connected";
    joinData.msg = nickname + " has joined the chat.";
    msgsCollection.addItem(joinData);
    // Attach the user's nickname to the socket
    socket.nickname = nickname;
    socket.broadcast.emit("notice", joinData);
    socket.emit("notice", joinData);
  });


  // When a user renames themself, send a notice
  // to all users except the originating client
  socket.on("rename", function (renameData) {
    var oldName = renameData.oldName;
    var newName = renameData.newName;

    console.log("rename: ", oldName, "to", newName);

    renameData.action = "user-renamed";
    renameData.msg = oldName + " has renamed themself " + newName;
    msgsCollection.addItem(renameData);
    // Attach the user's nickname to the socket
    socket.nickname = newName;
    socket.broadcast.emit("notice", renameData);
  });


  // 'query' expects an object with min and/or max timestamps and sends back
  // an array of messages which have a timestamp falling within that range
  socket.on("query", function (queryData) {
    var min  = queryData.minTimestamp || Number.MIN_SAFE_INTEGER;
    var max  = queryData.maxTimestamp || Number.MAX_SAFE_INTEGER;

    var items = getSortedItems();
    var res = [];

    for(var i = 0, size = items.length; i < size; i++) {
      var item = items[i];
      if(min < item.when && max > item.when) {
        res.push(item);
      }
    }

    console.log("query min: " + min + ", max: " + max + ", " + res.length + " matches of " + items.length);

    socket.emit("query", res);
  });


  // When a user disconnects, send a notice
  // to all users except the originating client
  socket.on("disconnect", function () {
    remove(users, socket);

    console.log("leave: ", socket.nickname, "(" + users.length + " users)");

    var leaveData = {
      action: "user-disconnect",
      msg: socket.nickname + " has left the chat.",
      nickname: socket.nickname,
      when: Date.now()
    };
    msgsCollection.addItem(leaveData);

    socket.broadcast.emit("notice", leaveData);
  });

});


var jokeCount = -1;
var curJoke = null;
var curCb = null;
var jokeState = null; // one of: ["knock", "who-there", "who", null]

function postJoke() {
  jokeCount++;

  setupAwaitTaskTimer(jokeIntervalMs, function (callback) {
    jokeCount++;
    jokeState = "knock";

    curJoke = jokes.jokes[Math.floor(jokes.jokes.length * Math.random())];
    curCb = callback;

    var msg = { msg: "Knock, knock.", when: Date.now() };
    emitToAll("notice", msg);
    msgsCollection.addItem(msg);
  });
}


// ==== utils ====

/** remove the first matching element from an array */
function remove(ary, val) {
  for (var i = 0, size = ary.length; i < size; i++) {
    if (ary[i] === val) {
      ary[i] = ary[size - 1];
      ary.pop();
      return true;
    }
  }
  return false;
}


/** emit a message to all listening sockets */
function emitToAll(type, data) {
  var clients = ioSocket.sockets.sockets;
  for (var i = 0, size = clients.length; i < size; i++) {
    clients[i].emit(type, data);
  }
}


/** run a task at a random time between now and maxIntervalMs and then wait for the task to return before setting up another timer to run it again and again... */
function setupAwaitTaskTimer(maxIntervalMs, task) {
  setTimeout(function () {
    var c = 0;
    task(function () {
      c++;
      if (c === 1) {
        setupAwaitTaskTimer(maxIntervalMs, task);
      }
    });
  }, Math.random() * maxIntervalMs);
}
